File: problem_4.rb
#Find the largest palindrome made from the product of two 3-digit numbers.
class LargestPalindrome

  private

  attr_accessor :factors_range

  def initialize(min_factor, max_factor)
    self.factors_range = (min_factor..max_factor)
  end

  def all_products
    factors_range.to_a.combination(2).map { |pair| pair[0] * pair[1] }.uniq
  end

  def is_palindrome(number)
    number.to_s == number.to_s.reverse
  end

  def all_palindromes
    all_products.select { |product| is_palindrome(product) }
  end

  public

  def largest_palindrome
    all_palindromes.max
  end

end

if __FILE__ == $PROGRAM_NAME
  puts LargestPalindrome.new(100, 999).largest_palindrome
end

File: problem_4.py
"""
Find the largest palindrome made from the product of two 3-digit numbers.
"""

import itertools

def main():
    print(largest(100, 999))

def all_products(min_factor, max_factor):
    return [pair[0]*pair[1] for pair in itertools.combinations(list(range(min_factor, max_factor)), 2)]

def is_palindrome(number):
    return str(number) == str(number)[::-1]

def all_palindromes(min_factor, max_factor):
    return [product for product in all_products(min_factor, max_factor) if
            is_palindrome(product)]

def largest(min_factor, max_factor):
    return max(all_palindromes(min_factor, max_factor))

if __name__ == "__main__":
    main()
